<a href="https://colab.research.google.com/github/CodeWithSridhar/Innomatics-Internship/blob/main/Task5_RegEx.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **Task - 5 (RegEx)**

**Questions Link**: https://www.hackerrank.com/domains/python?filters%5Bsubdomains%5D%5B%5D=py-regex 

## Detect Floating Point Number

In [ ]:
import re
class Main():
    def __init__(self):
        self.n = int(input())
        for i in range(self.n):
            self.s = input()
            print(bool(re.match(r'^[-+]?[0-9]*\.[0-9]+$', self.s)))
                    
if __name__ == '__main__':
    Answer = Main()

4
4.0O0
False
-1.00
True
+4.54
True
SomeRandomStuff
False


Explanation:

4.0O0 : O is not a digit.


-1.00 : is valid.

+4.54 : is valid.

SomeRandomStuff: is not a number.

## Group(), Groups() & Groupdict()

In [ ]:
import re

expression = r"([a-zA-Z0-9])\1+"
String = re.search(expression,input('Enter string : '))
if String:
    print(String.group(1))
else:
    print(-1)

Enter string : ..12345678910111213141516171820212223
1


Here, .. is the first repeating character, but it is not alphanumeric.
1 is the first (from left to right) alphanumeric repeating character of the string in the substring 111.

## Re.findall() & Re.finditer()

In [ ]:
import re
String = re.findall(r'(?<=[qwrtypsdfghjklzxcvbnm])([aeiou]{2,})(?=[qwrtypsdfghjklzxcvbnm])', input().strip(), re.IGNORECASE)

if String:
    for i in String:
        print(i)
else:
    print(-1)

rabcdeefgyYhFjkIoomnpOeorteeeeet
ee
Ioo
Oeo
eeeee


Explanation

ee is located between consonant d and f.

Ioo is located between consonant k and m.

Oeo is located between consonant p and r.

eeeee is located between consonant t and t.

## Re.start() & Re.end()

In [ ]:
import re

String, k = input(), input()
matches = re.finditer(r'(?=(' + k + '))', String)

Foundanymatch = False
for match in matches:
    Foundanymatch = True
    print((match.start(1), match.end(1) - 1))

if Foundanymatch == False:
    print((-1, -1))

aaadaa
aa
(0, 1)
(1, 2)
(4, 5)


## Regex Substitution

In [ ]:
import re

def Word(any_match):
    if any_match.group(1) == '&&':
        return 'and'
    else:
        return 'or'

for _ in range(int(input('Enter number of lines: '))):
    print(re.sub(r"(?<= )(\|\||&&)(?= )", Word,input('Enter string: ')))

Enter number of lines: 11
Enter string: a = 1;
a = 1;
Enter string: b = input();
b = input();
Enter string: 

Enter string: if a + b > 0 && a - b < 0:
if a + b > 0 and a - b < 0:
Enter string:  start()
 start()
Enter string: elif a*b > 10 || a/b < 1:
elif a*b > 10 or a/b < 1:
Enter string: stop()
stop()
Enter string: print set(list(a)) | set(list(b))
print set(list(a)) | set(list(b))
Enter string: #Note do not change &&& or ||| or & or |
#Note do not change &&& or ||| or & or |
Enter string: #Only change those '&&' which have space on both sides.
#Only change those '&&' which have space on both sides.
Enter string: #Only change those '|| which have space on both sides.
#Only change those '|| which have space on both sides.


## Validating phone numbers

In [ ]:
import re
N = int(input('Enter num of lines: '))

for i in range(N):
    Phone_number = input('Enter your phone number : ')
    if(len(Phone_number)==10 and Phone_number.isdigit()):
        Valid = re.findall(r"^[789]\d{9}$",Phone_number)
        if(len(Valid)==1):
            print("YES")
        else:
            print("NO")
    else:
        print("NO")

Enter num of lines: 2
Enter your phone number : 9587456281
YES
Enter your phone number : 1252478965
NO


## Validating Roman Numerals

You are given a string, and you have to validate whether it's a valid Roman numeral. If it is valid, print True. Otherwise, print False. Try to create a regular expression for a valid Roman numeral.

In [ ]:
regex_pattern = r""	# Do not delete 'r'.
Thousands = 'M{0,3}'
Hundreds = '(C[MD]|D?C{0,3})'
Tens = '(X[CL]|L?X{0,3})'
Ones = '(I[VX]|V?I{0,3})'
regex_pattern = r"%s%s%s%s$" % (Thousands, Hundreds, Tens, Ones)    

import re
print(str(bool(re.match(regex_pattern, input()))))

CDXXI
True


## Validating and Parsing Email Addresses

In [ ]:
import re
N = int(input('Enter number of lines : '))

for i in range(N):
    Name, E_mail = input('Enter Name and E-mail: ').split()
    Regular_pattern = "<[a-z][a-zA-Z0-9\-\.\_]+@[a-zA-Z]+\.[a-zA-Z]{1,3}>"
    if bool(re.match(Regular_pattern, E_mail)):
        print(Name,E_mail)

Enter number of lines : 2
Enter Name and E-mail: DEXTER <dexter@hotmail.com>
DEXTER <dexter@hotmail.com>
Enter Name and E-mail: VIRUS <virus!@variable.:p>


## Hex Color Code

In [ ]:
import re
Line = int(input())
in_css = False
for _ in range(Line):
    s = input()
    if '{' in s:
        in_css = True
    elif '}' in s:
        in_css = False
    elif in_css:
        for color in re.findall('#[0-9a-fA-F]{3,6}', s):
            print(color)

11
#BED
{
color: #FfFdF8; background-color:#aef;
#FfFdF8
#aef
font-size: 123px;
background: -webkit-linear-gradient(top, #f9f9f9, #fff);
#f9f9f9
#fff
}
#Cab
{
background-color: #ABC;
#ABC
border: 2px dashed #fff;
#fff
}


Explanation

    #BED and #Cab satisfy the Hex Color Code criteria, but they are used as selectors and not as color codes in the given CSS.

Hence, the valid color codes are:

    #FfFdF8
    #aef
    #f9f9f9
    #fff
    #ABC
    #fff

Note: There are no comments ( // or /* */) in CSS Code.

## Validating UID

In [ ]:
import re

for i in range(int(input())):
    N = input().strip()
    if N.isalnum() and len(N) == 10:
        if bool(re.search(r'(.*[A-Z]){2,}',N)) and bool(re.search(r'(.*[0-9]){3,}',N)):
            if re.search(r'.*(.).*\1+.*',N):
                print('Invalid')
            else:
                print('Valid')
        else:
            print('Invalid')
    else:
        print('Invalid')

2
Enter your UID : B1CD102354
Invalid
Enter your UID : B1CDEF2354
Valid


Explanation

B1CD102354:1 is repeating → Invalid

B1CDEF2354: Valid

## Validating Postal Codes

In [ ]:
regex_integer_in_range = r"^[1-9][\d]{5}$"	# Do not delete 'r'.
regex_alternating_repetitive_digit_pair = r"(\d)(?=\d\1)"	# Do not delete 'r'.


import re
P = input()

print (bool(re.match(regex_integer_in_range, P)) 
and len(re.findall(regex_alternating_repetitive_digit_pair, P)) < 2)

110000
False


Explanation 0

110000 : (0, 0) and (0, 0) are two alternating digit pairs.

Hence, it is an invalid postal code.

Note:
A score of 0 will be awarded for using 'if' conditions in your code.
You have to pass all the testcases to get a positive score.

## Detect HTML Tags, Attributes and Attribute Values

In [ ]:
from html.parser import HTMLParser
class MyHTMLParser(HTMLParser):
    def handle_starttag(self, tag, attrs):
        print(tag)
        for attr, value in attrs:
            print("->", attr, ">", value)

    def handle_startendtag(self, tag, attrs):
        print(tag)
        for attr, value in attrs:
            print("->", attr, ">", value)

Html = ''
for _ in range(int(input())):
    Html = Html + input().rstrip() + '\n'

parser = MyHTMLParser()
parser.feed(Html)
parser.close()

9
<head>
<title>HTML</title>
</head>
<object type="application/x-flash"
data="your-file.swf"
width="0" height="0">
<!-- <param name="movie" value="your-file.swf" /> -->
<param name="quality" value="high"/>
</object>
head
title
object
-> type > application/x-flash
-> data > your-file.swf
-> width > 0
-> height > 0
param
-> name > quality
-> value > high


Explanation:

    head tag: Print the head tag only because it has no attribute.

    title tag: Print the title tag only because it has no attribute.

    object tag: Print the object tag. In the next 4 lines, print the attributes type, data, width and height along with their respective values.

    <!-- Comment --> tag: Don't detect anything inside it.

    param tag: Print the param tag. In the next 2 lines, print the attributes name along with their respective values.

## HTML Parser - Part 1

In [ ]:
from html.parser import HTMLParser

class MyHTMLParser(HTMLParser):
    def handle_starttag(self, tag, attrs):
        print ('Start :', tag)
        for ele in attrs:
            print ('->', ele[0], '>', ele[1])

    def handle_endtag(self, tag):
        print ('End   :', tag)

    def handle_startendtag(self, tag, attrs):
        print ('Empty :', tag)
        for ele in attrs:
            print ('->', ele[0], '>', ele[1])

HTMLParser = MyHTMLParser()
for _ in range(int(input())):
    HTMLParser.feed(input())

2
<html><head><title>HTML Parser - I</title></head>
Start : html
Start : head
Start : title
End   : title
End   : head
<body data-modal-target class='1'><h1>HackerRank</h1><br /></body></html>
Start : body
-> data-modal-target > None
-> class > 1
Start : h1
End   : h1
Empty : br
End   : body
End   : html


## HTML Parser - Part 2

In [1]:
from html.parser import HTMLParser

class MyHTMLParser(HTMLParser):
    def handle_comment(self, data):
        if (len(data.split('\n')) != 1):
            print(">>> Multi-line Comment")
        else:
            print(">>> Single-line Comment")
        print(data.replace("\r", "\n"))
    def handle_data(self, data):
        if data.strip():
            print(">>> Data")
            print(data)
  
html = ""       
for i in range(int(input())):
    html += input().rstrip()
    html += '\n'
    
parser = MyHTMLParser()
parser.feed(html)
parser.close()

4
<!--[if IE 9]>IE9-specific content
<![endif]-->
<div> Welcome to HackerRank</div>
<!--[if IE 9]>IE9-specific content<![endif]-->
>>> Multi-line Comment
[if IE 9]>IE9-specific content
<![endif]
>>> Data
 Welcome to HackerRank
>>> Single-line Comment
[if IE 9]>IE9-specific content<![endif]


## Validating Credit Card Numbers

In [2]:
import re
for i in range(int(input())):
    S = input().strip()
    pre_match = re.search(r'^[456]\d{3}(-?)\d{4}\1\d{4}\1\d{4}$',S)
    if pre_match:
        processed_string = "".join(pre_match.group(0).split('-'))
        final_match = re.search(r'(\d)\1{3,}',processed_string)
        if final_match:
            print('Invalid')
        else :
            print('Valid')
    else:
        print('Invalid')

6
4123456789123456
Valid
5123-4567-8912-3456
Valid
61234-567-8912-3456
Invalid
4123356789123456
Valid
5133-3367-8912-3456
Invalid
5123 - 3567 - 8912 - 3456
Invalid
